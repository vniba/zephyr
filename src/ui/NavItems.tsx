import {
  BOOKINGS,
  CABINS,
  DASHBOARD,
  SETTINGS,
  USERS,
} from '../utils/constants.ts';
import { HiHome, HiMiniHomeModern, HiUsers } from 'react-icons/hi2';
import { FaCalendarAlt } from 'react-icons/fa';
import { MdSettings } from 'react-icons/md';

export const navItems = [
  {
    to: DASHBOARD,
    icon: <HiHome />,
    label: 'Home',
  },
  {
    to: BOOKINGS,
    icon: <FaCalendarAlt />,
    label: 'Bookings',
  },
  {
    to: CABINS,
    icon: <HiMiniHomeModern />,
    label: 'Cabins',
  },
  {
    to: USERS,
    icon: <HiUsers />,
    label: 'Users',
  },
  {
    to: SETTINGS,
    icon: <MdSettings />,
    label: 'Settings',
  },
];
