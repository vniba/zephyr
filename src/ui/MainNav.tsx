import styled from 'styled-components';
import { NavLink } from 'react-router-dom';
import { navItems } from './NavItems.tsx';

const NavList = styled.ul`
  display: flex;
  flex-direction: column;
  gap: 0.8rem;
`;

const StyledNavLink = styled(NavLink)`
  &:link,
  &:visited {
    display: flex;
    align-items: center;
    gap: 1.2rem;

    color: var(--color-grey-600);
    font-size: 1.6rem;
    font-weight: 500;
    padding: 1.2rem 2.4rem;
    transition: all 0.3s;
  }

  /* This works because react-router places the active class on the active NavLink */
  &:hover,
  &:active,
  &.active:link,
  &.active:visited {
    color: var(--color-grey-700);
    background-color: var(--color-grey-50);
    border-radius: var(--border-radius-sm);
  }

  & svg {
    width: 2.4rem;
    height: 2.4rem;
    color: var(--color-grey-400);
    transition: all 0.3s;
  }

  &:hover svg,
  &:active svg,
  &.active:link svg,
  &.active:visited svg {
    color: var(--color-brand-900);
  }
`;

function MainNav() {
  return (
    <nav>
      <NavList>
        {navItems.map(item => (
          <li key={item.label}>
            <StyledNavLink to={item.to}>
              {item.icon}
              <span>{item.label}</span>
            </StyledNavLink>
          </li>
        ))}
      </NavList>
    </nav>
  );
}

export default MainNav;
