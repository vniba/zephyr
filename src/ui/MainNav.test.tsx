import { render, screen } from '@testing-library/react';
import MainNav from './MainNav.tsx';
import { expect, it } from 'vitest';
import { BrowserRouter } from 'react-router-dom';
import { navItems } from './NavItems.tsx';

describe('MainNav component', function () {
  it('should render  li tag with length of navItems ', function () {
    render(
      <BrowserRouter>
        <MainNav />
      </BrowserRouter>,
    );
    const list = screen.getAllByRole('listitem');
    expect(list).toHaveLength(navItems.length);
  });
  it('should render href with correct path', function () {
    render(
      <BrowserRouter>
        <MainNav />
      </BrowserRouter>,
    );

    navItems.forEach(item => {
      const linkEl = screen.getByRole('link', { name: item.label });
      expect(linkEl).toHaveAttribute('href', `/${item.to}`);
    });
  });

  it('should render icons with length of navItem', function () {
    const { container } = render(
      <BrowserRouter>
        <MainNav />
      </BrowserRouter>,
    );
    const svg = container.querySelectorAll('svg');
    expect(svg).toHaveLength(navItems.length);
  });

  it('should match default snapshot', function () {
    const { container } = render(
      <BrowserRouter>
        <MainNav />
      </BrowserRouter>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
