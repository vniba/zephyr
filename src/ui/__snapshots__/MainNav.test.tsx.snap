// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MainNav component > should match default snapshot 1`] = `
.c0 {
  display: flex;
  flex-direction: column;
  gap: 0.8rem;
}

.c1:link,
.c1:visited {
  display: flex;
  align-items: center;
  gap: 1.2rem;
  color: var(--color-grey-600);
  font-size: 1.6rem;
  font-weight: 500;
  padding: 1.2rem 2.4rem;
  transition: all 0.3s;
}

.c1:hover,
.c1:active,
.c1.active:link,
.c1.active:visited {
  color: var(--color-grey-700);
  background-color: var(--color-grey-50);
  border-radius: var(--border-radius-sm);
}

.c1 svg {
  width: 2.4rem;
  height: 2.4rem;
  color: var(--color-grey-400);
  transition: all 0.3s;
}

.c1:hover svg,
.c1:active svg,
.c1.active:link svg,
.c1.active:visited svg {
  color: var(--color-brand-900);
}

<nav>
  <ul
    class="c0"
  >
    <li>
      <a
        class="c1"
        href="/dashboard"
      >
        <svg
          aria-hidden="true"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M11.47 3.84a.75.75 0 011.06 0l8.69 8.69a.75.75 0 101.06-1.06l-8.689-8.69a2.25 2.25 0 00-3.182 0l-8.69 8.69a.75.75 0 001.061 1.06l8.69-8.69z"
          />
          <path
            d="M12 5.432l8.159 8.159c.03.03.06.058.091.086v6.198c0 1.035-.84 1.875-1.875 1.875H15a.75.75 0 01-.75-.75v-4.5a.75.75 0 00-.75-.75h-3a.75.75 0 00-.75.75V21a.75.75 0 01-.75.75H5.625a1.875 1.875 0 01-1.875-1.875v-6.198a2.29 2.29 0 00.091-.086L12 5.43z"
          />
        </svg>
        <span>
          Home
        </span>
      </a>
    </li>
    <li>
      <a
        class="c1"
        href="/bookings"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 448 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 464c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V192H0v272zm320-196c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v40c0 6.6-5.4 12-12 12h-40c-6.6 0-12-5.4-12-12v-40zm0 128c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v40c0 6.6-5.4 12-12 12h-40c-6.6 0-12-5.4-12-12v-40zM192 268c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v40c0 6.6-5.4 12-12 12h-40c-6.6 0-12-5.4-12-12v-40zm0 128c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v40c0 6.6-5.4 12-12 12h-40c-6.6 0-12-5.4-12-12v-40zM64 268c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v40c0 6.6-5.4 12-12 12H76c-6.6 0-12-5.4-12-12v-40zm0 128c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v40c0 6.6-5.4 12-12 12H76c-6.6 0-12-5.4-12-12v-40zM400 64h-48V16c0-8.8-7.2-16-16-16h-32c-8.8 0-16 7.2-16 16v48H160V16c0-8.8-7.2-16-16-16h-32c-8.8 0-16 7.2-16 16v48H48C21.5 64 0 85.5 0 112v48h448v-48c0-26.5-21.5-48-48-48z"
          />
        </svg>
        <span>
          Bookings
        </span>
      </a>
    </li>
    <li>
      <a
        class="c1"
        href="/cabins"
      >
        <svg
          aria-hidden="true"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 20 20"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M14.916 2.404a.75.75 0 01-.32 1.012l-.596.31V17a1 1 0 01-1 1h-2.26a.75.75 0 01-.75-.75v-3.5a.75.75 0 00-.75-.75H6.75a.75.75 0 00-.75.75v3.5a.75.75 0 01-.75.75h-3.5a.75.75 0 010-1.5H2V9.957a.75.75 0 01-.596-1.372L2 8.275V5.75a.75.75 0 011.5 0v1.745l10.404-5.41a.75.75 0 011.012.32zM15.861 8.57a.75.75 0 01.736-.025l1.999 1.04A.75.75 0 0118 10.957V16.5h.25a.75.75 0 110 1.5h-2a.75.75 0 01-.75-.75V9.21a.75.75 0 01.361-.64z"
          />
        </svg>
        <span>
          Cabins
        </span>
      </a>
    </li>
    <li>
      <a
        class="c1"
        href="/users"
      >
        <svg
          aria-hidden="true"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M4.5 6.375a4.125 4.125 0 118.25 0 4.125 4.125 0 01-8.25 0zM14.25 8.625a3.375 3.375 0 116.75 0 3.375 3.375 0 01-6.75 0zM1.5 19.125a7.125 7.125 0 0114.25 0v.003l-.001.119a.75.75 0 01-.363.63 13.067 13.067 0 01-6.761 1.873c-2.472 0-4.786-.684-6.76-1.873a.75.75 0 01-.364-.63l-.001-.122zM17.25 19.128l-.001.144a2.25 2.25 0 01-.233.96 10.088 10.088 0 005.06-1.01.75.75 0 00.42-.643 4.875 4.875 0 00-6.957-4.611 8.586 8.586 0 011.71 5.157v.003z"
          />
        </svg>
        <span>
          Users
        </span>
      </a>
    </li>
    <li>
      <a
        class="c1"
        href="/settings"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0V0z"
            fill="none"
          />
          <path
            d="M19.14 12.94c.04-.3.06-.61.06-.94 0-.32-.02-.64-.07-.94l2.03-1.58a.49.49 0 00.12-.61l-1.92-3.32a.488.488 0 00-.59-.22l-2.39.96c-.5-.38-1.03-.7-1.62-.94l-.36-2.54a.484.484 0 00-.48-.41h-3.84c-.24 0-.43.17-.47.41l-.36 2.54c-.59.24-1.13.57-1.62.94l-2.39-.96c-.22-.08-.47 0-.59.22L2.74 8.87c-.12.21-.08.47.12.61l2.03 1.58c-.05.3-.09.63-.09.94s.02.64.07.94l-2.03 1.58a.49.49 0 00-.12.61l1.92 3.32c.12.22.37.29.59.22l2.39-.96c.5.38 1.03.7 1.62.94l.36 2.54c.05.24.24.41.48.41h3.84c.24 0 .44-.17.47-.41l.36-2.54c.59-.24 1.13-.56 1.62-.94l2.39.96c.22.08.47 0 .59-.22l1.92-3.32c.12-.22.07-.47-.12-.61l-2.01-1.58zM12 15.6c-1.98 0-3.6-1.62-3.6-3.6s1.62-3.6 3.6-3.6 3.6 1.62 3.6 3.6-1.62 3.6-3.6 3.6z"
          />
        </svg>
        <span>
          Settings
        </span>
      </a>
    </li>
  </ul>
</nav>
`;
